htcanalyze/htcanalyze.py

```diff
@@ -65,7 +65,13 @@ def manage_thresholds(self, resources: dict) -> dict:
         The important part is that the keywords exist
         "Usage", "Requested"
 
-        :param resources: resource dictionary
+        :param resources: a dict like:
```

Comment:
I suggest to make a resource class for this.

```
class Resource(object):
    def __init__(resources: str, usage: int, requested: int, allocated: int):
        self.name = resources
        self.usage = usage
        self.requested = requested
        self.allocated = allocated
        self.usage_notification = None
        self.usage_level = None


def refactor_resources(resources: dict) -> List[Resource]:
    """Convert a dict of lists to a list of Resource objects"""
    resources = {k.lower(): v for k, v in resources.items()}
    resources = [dict(zip(resources, v)) for v in zip(*resources.values())]
    resources = [Resource(**resource) for resource in resources]
    return resources
```

Instead of passing a dictionary object to the class, you pass a list of Resource objects to the class (or, alternatively, you convert the dictionary to a list resource objects inside of this class -- whichever fits better logistically).

Then instead of `n_resources = len(resources['Resources'])` you can do the following:

```
for resource in resources:
    if resource.requested != 0:
        deviation = resource.usage / resource.requested  # note: they become floats, no need to pre-assign them
        
        if 1 - self.bad_usage <= deviation <= 1 + self.bad_usage:
            resource.usage_level = 'error'
        elif 1 - self.tolerated_usage <= deviation <= 1 + self.tolerated_usage:        
            resource.usage_level = 'warning'
        elif str(usage) == 'nan':
            resource.usage_level = 'light_warning'  # or some other name to indicate usage of 'yellow2'
        else:
            resource.usage_level = 'normal'
```

In a later part of the code (the part that is all about providing the text output, separated from this level of logic), you can then have:

```
def get_color(level: str) -> str:
    """Convert an alert level to an appropriate color"""
    colors = {'error': 'red', 'warning': 'yellow', 'light_warning': 'yellow2', 'normal': 'green'}
    return colors[level]

def print_output(resources: List[Resource]) -> None:
    """This is just a dummy function to explain the printing side of things"""
    for resource in resources:
        color = get_color(resource.usage_level)
        print(f'[{color}]{resource.usage}[/{color}]')
```

Reply:
I like this idea 👍

Reply:
This Resource class should definitely be declared in an own file and the logvllidator should create the instances, when reading the log files. This idea could be extended to more dictionaries, like the time_dict or the occurred_errors dict, but I wouldn't focus too much on this. 
I really like the idea of the resource dict, cause this gets shovelled around pretty often.

Reply:
It's a good idea to provide clarity to the format of the input you're getting. Now we know exactly what a resource looks like. If there are more "objects" like this, it's a good idea to make classes for them. I haven't delved too deeply into this, but you can also use getter and setter methods to change instance arguments.

Reply:
Create Issue for this idea #118, this should have it's own PR.